Apply multiple stability fixes and release a new qTox version

nickolay168 opened the pull request:
I have applied some stability fixes, here is the list from my change log:

* Show light object to camera while speaking, the application will fail (Fixed in [toxcore](https://github.com/TokTok/c-toxcore/pull/2768) ETA in v. 0.2.20).
* Fix the failure when logging is switched off and user tries to get detail for the friend.
* SIGABRT during the attempt to delete a friend.
* Setting NoSpam does not have an effect.

I have decided also to bump the version, to avoid issues with the same versions with different code.

nickolay168 committed:
Add patch with multiple stability fixes.

kwizart commented:
If you are interested in this package. Feel free to create an account and help co-maintain it.
https://rpmfusion.org/Contributors

nickolay168 commented:
> If you are interested in this package. Feel free to create an account and help co-maintain it. https://rpmfusion.org/Contributors

Sure, I have followed the instructions on the page. @kwizart what should be the next steps?

kwizart commented:
@nickolay168 Usually it should be the current maintainer to ack.
Can you create a ticket in our bugtracker at bugzilla.rpmfusion.org.
Then ask the maintainer if you can co-maintain this package.

Unless @eclipseo can answer here ?

nickolay168 commented:
Thank you!
@kwizart, I have created an issue in bugzilla: https://bugzilla.rpmfusion.org/show_bug.cgi?id=7123

sergiomb2 commented:
Hi, sorry for delay,  since you already did https://github.com/nickolay168/qTox/releases/tag/v1.17.7 , I prefer use the tag ,
to not hack qtox.spec in all sections

nickolay168 committed:
Use the new repository

nickolay168 commented:
> Hi, sorry for delay, since you already did https://github.com/nickolay168/qTox/releases/tag/v1.17.7 , I prefer use the tag , to not hack qtox.spec in all sections

Sure, Thank you! I have updated the PR. I have also used a new tag for version, where I have removed unsupported unfinished code.

sergiomb2 commented:
I fixed two minor details and squash commits , push and built for Fedora 41 and rawhide .
Should we build also for F40 ?

eclipseo commented:
I ain't sure about this commit for ffmpeg 7.x

https://github.com/qTox/qTox/commit/d70e9a8d5daeae0f168a8d8640bb6edae7dc33d4

Could you explain why you choose to remove avformat_free_context() ?

Could you fix this alignment:

https://github.com/qTox/qTox/commit/c64d79fa8e078c2abacf0bdf8c87e818b22673ad#diff-e8b9f765aa2e392d5753e626bcb07aa1a8e78346da63be1ccf361ad402e280bdR316

I know it's not Python, but still :smile: 

Could you check why you have absolute paths in translations.qrc:

https://github.com/qTox/qTox/commit/3f3a7e641496959507a884d3b3aedfd81fd23df4#diff-4233401a9e85a25235756d3625491988bf9b6294605080284bb435698dc6c542R1

eclipseo commented:
Also, ok this is an unofficial fork, I've checked your changes, and there are no glaring issues.

But there are also unofficial forks like : https://github.com/TokTok/qTox and they have @iphydf on board who is one of the main c-toxcore dev.

Their fork is very active, with a lot more changes, and with more "stars".

I am more inclined to use their fork as the new basis rather than yours, because you seem to work solo. They have applied a similar solution for security by also removing the Extension message support https://github.com/TokTok/qTox/commit/f9d229ff111a124aee023e919e9a5a7e26845f87

Considering what happen last year with xz utils, considering the sensibility of this project, I'd rather be cautious and use the fork made by people working also on c-toxcore.

Would you have a problem contributing your change to their repo and working with them?

cc @sergiomb2 @kwizart for opinions.

iphydf commented on qTox-1.17.7.patch:
```diff
@@ -0,0 +1,408 @@
+     std::atomic<IAudioControl*> audio;
++    // atomic flag showing that we do not need to accept frames as the cancel
++    // call request was sent.
++    std::atomic<bool> isCancelling;
```
I think this is a fix for a crash/race condition in closing qTox during a call. We've made a few changes (e.g. https://github.com/TokTok/qTox/pull/274 and https://github.com/TokTok/qTox/pull/192) towards that. Can you check whether this change is still necessary?

nickolay168 replied:
It seems like it is still needed, because it solves different issue. It is fixing rare situation, when user tryes to cancel the call AND simultaneously the Video/Audio frame comes in. If I am remembering it correctly that is what happens 
1. User hits hang up and the  `QWriteLocker locker{&callsLock};` will lock the frame frocessing
2. Now in very edge case, the frame comes between the (1) and the call `toxav_call_control(toxav.get(), friendNum, TOXAV_CALL_CONTROL_CANCEL, &err);` which will never complete as it will hit the mutex inside toxcore (`ToxAv->mutex`). This mutex will wait for frame to be processed , which will never happen because of (1)

[#274](https://github.com/TokTok/qTox/pull/274) Fixed the lambda expression, sending frames, when the call is being terminated.
[#192](https://github.com/TokTok/qTox/pull/192) Makes sure that the calls are being cleaned up when `CoreAV` is being destroyed.

So it seems that is a separate issue, which is an edge case, but still can happen.

iphydf commented on qTox-1.17.7.patch:
```diff
@@ -0,0 +1,408 @@
++#ifdef SPELL_CHECKING
++    if (decorator != nullptr) {
++    	delete decorator;
++    	decorator = nullptr;
```
https://github.com/TokTok/qTox/pull/159

iphydf commented on qTox-1.17.7.patch:
```diff
@@ -0,0 +1,408 @@
+     }
+ 
+-    for (auto form : chatForms) {
+-        delete form;
```
This change makes sense (given QObject child deletion), but what issue does it fix? Does deleting these prematurely cause a problem?

nickolay168 replied:
This code does not causes issues by itself, in this change I just wanted to unify  handling of `chatForm` and `groupChatForm`.

sergiomb2 commented:
https://github.com/nickolay168/qtox tag v1.17.8 is what build on RPMFusion . 
https://github.com/TokTok/qTox have a lot more commits @nickolay168  what do you think , can you merge your work in the TokTok repo ? 
Thanks and best regards,

iphydf commented:
@nickolay168 648BF2EEE794E94444B848F8FC6AD3BA029C9BC2649BA761EF556DA17F549022A8D7596E7DBA is a groupbot that will invite you to the toktok-dev chat. I think there are at least a few fixes in your repo that we didn't address yet (e.g. bugs when history is off, because I've never tested that). Let's get them merged :).

nickolay168 commented:
I have also a PR to port qTox to KDE6, but I was testing it. Let me look 
tomorrow, how we can work it out.

Thank you!


                            Nikolay Rovinskiy

> @nickolay168 648BF2EEE794E94444B848F8FC6AD3BA029C9BC2649BA761EF556DA17F549022A8D7596E7DBA is a groupbot that will invite you to the toktok-dev chat. I think there are at least a few fixes in your repo that we didn't address yet (e.g. bugs when history is off, because I've never tested that). Let's get them merged :).
>

iphydf commented:
That's also done. We're on Qt6 now.

nickolay168 commented:
> I ain't sure about this commit for ffmpeg 7.x
> 
> [qTox/qTox@d70e9a8](https://github.com/qTox/qTox/commit/d70e9a8d5daeae0f168a8d8640bb6edae7dc33d4)
> 
> Could you explain why you choose to remove avformat_free_context() ?
> 
> Could you fix this alignment:
> 
> [qTox/qTox@c64d79f#diff-e8b9f765aa2e392d5753e626bcb07aa1a8e78346da63be1ccf361ad402e280bdR316](https://github.com/qTox/qTox/commit/c64d79fa8e078c2abacf0bdf8c87e818b22673ad#diff-e8b9f765aa2e392d5753e626bcb07aa1a8e78346da63be1ccf361ad402e280bdR316)
> 
> I know it's not Python, but still 😄
> 
> Could you check why you have absolute paths in translations.qrc:
> 
> [qTox/qTox@3f3a7e6#diff-4233401a9e85a25235756d3625491988bf9b6294605080284bb435698dc6c542R1](https://github.com/qTox/qTox/commit/3f3a7e641496959507a884d3b3aedfd81fd23df4#diff-4233401a9e85a25235756d3625491988bf9b6294605080284bb435698dc6c542R1)

The official qTox version (https://github.com/qTox/qTox) used to take the list of devices by manually creating the AVFormatContext, which in turn involved copying the structure of AVInputFormat's AVClass that required allocation of private data for the specific AVClass, contained  in the iformat variable. The manual allocation required the size of this class, however, the priv_data_size field of AVInputFormat was long deprecated and finally removed in ffmpeg v. 7. Here is the commit in ffmpeg repository, which has broken qTox: 
https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/b800327f4c7233d09baca958121722a04c2035ff
 
and, specifically, the patch file to avformat.h, which used to define priv_data_size: 
https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/b800327f4c7233d09baca958121722a04c2035ff
In the commit I am using avdevice_list_input_sources function, which does not require  AVFormatContext, so it is not created and is not deleted :). I just have forgot to remove the commented code.

nickolay168 commented:
> https://github.com/nickolay168/qtox tag v1.17.8 is what build on RPMFusion . https://github.com/TokTok/qTox have a lot more commits @nickolay168 what do you think , can you merge your work in the TokTok repo ? Thanks and best regards,

Yes, probably it is the best way to go forward.

sergiomb2 commented:
> > https://github.com/nickolay168/qtox tag v1.17.8 is what build on RPMFusion . https://github.com/TokTok/qTox have a lot more commits @nickolay168 what do you think , can you merge your work in the TokTok repo ? Thanks and best regards,
> 
> Yes, probably it is the best way to go forward.

so also approve start using https://github.com/TokTok/qTox ?

nickolay168 commented:
> > > https://github.com/nickolay168/qtox tag v1.17.8 is what build on RPMFusion . https://github.com/TokTok/qTox have a lot more commits @nickolay168 what do you think , can you merge your work in the TokTok repo ? Thanks and best regards,
> > 
> > 
> > Yes, probably it is the best way to go forward.
> 
> so also approve start using https://github.com/TokTok/qTox ?

I Think, generally yes, as it seems to be the next official qTox repo.
@iphydf do you know if the new version works on libav > 6.0? It seems [here](https://github.com/TokTok/qTox/blob/464939aff9163d7e10e77af9c49659490f2821e2/src/video/cameradevice.cpp#L349) we are skipping copying of  private data. Does the device listing work without it? I remember some issues when these data were not there.

iphydf commented:
TokTok/qTox is built against ffmpeg 7.1: https://github.com/TokTok/dockerfiles/blob/master/qtox/download/download_ffmpeg.sh#L10

We also build against some older versions on CI. Whichever is in Ubuntu 22.04 or debian stable.

nickolay168 commented:
> TokTok/qTox is built against ffmpeg 7.1: https://github.com/TokTok/dockerfiles/blob/master/qtox/download/download_ffmpeg.sh#L10
> 
> We also build against some older versions on CI. Whichever is in Ubuntu 22.04 or debian stable.

@sergiomb2 Then it is totally reasonable to move to the new branch. It would be also great to use toxcore 0.2.20. It looks like it contains some great fixes.

sergiomb2 commented:
ok ,  qtox update to 1.18.3  on koji , should be published in a couple of days 

[qtox-1.18.3-1.fc40](https://koji.rpmfusion.org/koji/buildinfo?buildID=31443) 	[sergiomb](https://koji.rpmfusion.org/koji/userinfo?userID=37) 	2025-03-23 19:39:36 	complete
[qtox-1.18.3-1.fc41](https://koji.rpmfusion.org/koji/buildinfo?buildID=31442) 	[sergiomb](https://koji.rpmfusion.org/koji/userinfo?userID=37) 	2025-03-23 19:19:10 	complete
[qtox-1.18.3-1.fc42](https://koji.rpmfusion.org/koji/buildinfo?buildID=31439) 	[sergiomb](https://koji.rpmfusion.org/koji/userinfo?userID=37) 	2025-03-23 04:35:27 	complete
[qtox-1.18.3-1.fc43](https://koji.rpmfusion.org/koji/buildinfo?buildID=31438) 	[sergiomb](https://koji.rpmfusion.org/koji/userinfo?userID=37) 	2025-03-23 04:04:01 	complete
